src/compiler/compiler.rs

```diff
@@ -921,129 +944,176 @@ where
     T: CommandCreatorSync,
 {
     trace!("detect_compiler: {}", executable.display());
-
     // First, see if this looks like rustc.
 
-    let rustc_executable = if is_rustc_like(executable) {
-        Some(executable.to_path_buf())
+    let mut must_be_rust = false;
+    let mut rustc_executable = executable.to_path_buf();
+    if is_rustc_like(executable) {
+        must_be_rust = true;
     } else if env.iter().any(|(k, _)| k == OsStr::new("CARGO")) {
         // If not, detect the scenario where cargo is configured to wrap rustc with something other than sccache.
         // This happens when sccache is used as a `RUSTC_WRAPPER` and another tool is used as a
         // `RUSTC_WORKSPACE_WRAPPER`. In that case rustc will be the first argument rather than the command.
         //
         // The check for the `CARGO` env acts as a guardrail against false positives.
         // https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-reads
-        args.iter().next().and_then(|arg1| {
+        if let Some(arg1) = args.iter().next() {
             if is_rustc_like(arg1) {
-                Some(PathBuf::from(arg1))
-            } else {
-                None
+                must_be_rust = true;
+                rustc_executable = PathBuf::from(arg1)
             }
-        })
-    } else {
-        None
+        }
```

Comment:
I'd prefer to use a local `struct X {}` and avoid too many `mut` variables floating around in a scope.